src/python/pants/core/goals/deploy.py

```diff
@@ -115,51 +128,300 @@ async def publish_process_for_target(
+                    if self.outcome == _DeployStepOutcome.FAILURE
+                    else "failed dependencies"
+                )
+                prep = "for"
+
+            if self.description:
+                status += f" {prep} {self.description}"
+
+        result = []
+        for name in self.names:
+            result.append(f"{name} {status}")
+
+        return tuple(result)
+
+    @property
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for substep in self.substeps:
+            yield from substep.closure
+
+        yield self
+
+
+class _DeployStepResults(Collection[_DeployStepResult]):
+    @memoized_property
+    def exit_code(self) -> int:
+        if len(self) == 0:
+            return 0
+
+        return max(result.exit_code for result in self)
+
+    @property
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for result in self:
+            yield from result.closure
+
+
+@rule(desc="Publish package dependencies", level=LogLevel.DEBUG)
+async def _run_publish_process(publish: PublishPackages) -> _DeployStepResult:
+    if publish.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, publish.process.process)
+        return _DeployStepResult.from_publish_process_result(result, publish)
+
+    return _DeployStepResult.skipped(
+        action=_DeployStepAction.PUBLISH_PACKAGE,
+        names=publish.names,
+        description=publish.description,
+    )
+
+
+@rule(desc="Run deployment step", level=LogLevel.DEBUG)
+async def _run_deploy_process(deploy: DeployProcess) -> _DeployStepResult:
+    publish_results: Iterable[_DeployStepResult] = []
+    if deploy.publish_dependencies:
+        publish_processes = await _publish_processes_for_targets(deploy.publish_dependencies)
+        publish_results = _DeployStepResults(
+            await MultiGet(
+                Get(_DeployStepResult, PublishPackages, process) for process in publish_processes
+            )
+        )
+
+        if publish_results.exit_code != 0:
+            return _DeployStepResult.failed_dependencies(deploy, publish_results)
+
+    if deploy.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, deploy.process.process)
+        return _DeployStepResult.from_deploy_process_result(
+            result,
+            deploy,
+            substeps=publish_results,
+        )
+
+    return _DeployStepResult.skipped(
+        action=_DeployStepAction.DEPLOYMENT,
+        names=(deploy.name,),
+        description=deploy.description,
+        substeps=publish_results,
+    )
+
+
+@rule(desc="Deploy target dependencies", level=LogLevel.DEBUG)
+async def _deploy_target_dependencies(
+    request: _DeployTargetDependenciesRequest,
+) -> _DeployTargetRequests:
+    def should_ignore(dep: CoarsenedTarget) -> bool:
```

Comment:
Why do we need this? When could us == them?

And why not just compare the addresses directly instead of just the spec_path and target_name?

Reply:
that's indeed a good point, took this from another place where similar check was done to build a class path graph but it's true that it is a bit pointless.

Reply:
And to clarify, why do we need this check at all?

Reply:
This makes me a little uncomfortable. Are we sure we understand this logic? Maybe it is necessary!

Reply:
I was exploring this logic when trying to implement making the deploy process walk down the tree of dependencies. I had started with a rule that had a `_DeployTargetDependenciesRequest -> _DeployStepResults` signature that performed the recursive step and I was experiencing an infinite loop in Pants.

I added the conditional check when comparing my implementation with the one used in the JVM and thought that maybe `CoarsenedTarget` includes the representative target in its `dependencies` - as that would explain the infinite loop. It initially didn't solve the infinite loop problem and while I was debugging I could notice that my assumption was wrong: `dependencies` only included the actual dependencies.

The solution to the infinite loop was to break that _recursive_ rule `_DeployTargetDependenciesRequest -> _DeployStepResults` into two rules: `_DeployTargetDependenciesRequest -> _DeployTargetRequests` and `_DeployTargetRequests -> _DeployStepResults`. The conditional we are talking about here was a leftover artifact.